test/core/transport/call_filters_test.cc

```diff
@@ -0,0 +1,1466 @@
+  Filter2 f2;
+  d.AddFilter(&f2);
+  d.AddFilter(&f1);
+  EXPECT_EQ(d.call_data_alignment, alignof(void*));
+  // No padding needed, so just the sum of sizes.
+  EXPECT_EQ(d.call_data_size, sizeof(void*) + 1);
+  EXPECT_EQ(d.filters.size(), 2u);
+  // Check channel data
+  EXPECT_EQ(d.filters[0].channel_data, &f2);
+  EXPECT_EQ(d.filters[1].channel_data, &f1);
+  // Check call offsets
+  EXPECT_EQ(d.filters[0].call_offset, 0);
+  EXPECT_EQ(d.filters[1].call_offset, sizeof(void*));
+}
+
+TEST(StackDataTest, EmptyFilter) {
+  struct Filter1 {
+    struct Call {};
+  };
+  static_assert(std::is_empty<typename Filter1::Call>::value,
+                "Expect empty filter");
+  StackData d;
+  Filter1 f1;
+  d.AddFilter(&f1);
+  EXPECT_EQ(d.call_data_size, 0);
+}
+
+TEST(StackDataTest, OneFilterThenManyEmptyThenOneNonEmpty) {
+  struct Filter1 {
+    struct Call {
+      char c;
+    };
+  };
+  struct Filter2 {
+    struct Call {};
+  };
+  StackData d;
+  // Declare filters
+  Filter1 f1a;
+  Filter2 f2a;
+  Filter2 f2b;
+  Filter2 f2c;
+  Filter2 f2d;
+  Filter1 f1b;
+  // Add filters
+  d.AddFilter(&f1a);
+  d.AddFilter(&f2a);
+  d.AddFilter(&f2b);
+  d.AddFilter(&f2c);
+  d.AddFilter(&f2d);
+  d.AddFilter(&f1b);
+  // Check overall size
+  EXPECT_EQ(d.call_data_size, 2);
+  EXPECT_EQ(d.filters.size(), 6u);
+  // Check channel data
+  EXPECT_EQ(d.filters[0].channel_data, &f1a);
+  EXPECT_EQ(d.filters[1].channel_data, &f2a);
+  EXPECT_EQ(d.filters[2].channel_data, &f2b);
+  EXPECT_EQ(d.filters[3].channel_data, &f2c);
+  EXPECT_EQ(d.filters[4].channel_data, &f2d);
+  EXPECT_EQ(d.filters[5].channel_data, &f1b);
+  // Check call offsets
+  EXPECT_EQ(d.filters[0].call_offset, 0);
+  EXPECT_EQ(d.filters[1].call_offset, 0);
+  EXPECT_EQ(d.filters[2].call_offset, 0);
+  EXPECT_EQ(d.filters[3].call_offset, 0);
+  EXPECT_EQ(d.filters[4].call_offset, 0);
+  EXPECT_EQ(d.filters[5].call_offset, 1);
+}
+
+TEST(StackDataTest, FilterInit) {
+  struct Filter1 {
+    struct Call {
+      std::unique_ptr<int> p{new int(42)};
+    };
+  };
+  StackData d;
+  Filter1 f1;
+  d.AddFilter(&f1);
+  EXPECT_EQ(d.filters.size(), 1u);
+  EXPECT_EQ(d.filters[0].channel_data, &f1);
+  EXPECT_EQ(d.filters[0].call_offset, 0);
+  void* p = gpr_malloc_aligned(d.call_data_size, d.call_data_alignment);
+  d.filters[0].call_init(p, &f1);
+  EXPECT_EQ(*static_cast<Filter1::Call*>(p)->p, 42);
+  d.filters[0].call_destroy(p);
+  gpr_free_aligned(p);
+}
+
+TEST(StackDataTest, FilterInitWithArg) {
+  struct Filter1 {
+    struct Call {
+      explicit Call(Filter1* f) : p(new Filter1*(f)) {}
+      std::unique_ptr<Filter1*> p;
+    };
+  };
+  StackData d;
+  Filter1 f1;
+  d.AddFilter(&f1);
+  EXPECT_EQ(d.filters.size(), 1u);
+  EXPECT_EQ(d.filters[0].channel_data, &f1);
+  EXPECT_EQ(d.filters[0].call_offset, 0);
+  void* p = gpr_malloc_aligned(d.call_data_size, d.call_data_alignment);
+  d.filters[0].call_init(p, &f1);
+  EXPECT_EQ(*static_cast<Filter1::Call*>(p)->p, &f1);
```

Comment:
Does this actually do anything, given that the filter has no data members?

Reply:
It calls the constructor... I think I'm going to optimize this away in a followup... but consider:

```
class Filter {
  class Call {
    Call() { printf("call created\n");
    ~Call() { printf("call destroyed\n");
  };
};
```

It would be incorrect to omit these constructors/destructors since they have side effects.

If we can however prove that there are no side effects, then we can omit them.